### PR TITLE
implemented create strategy function

### PR DESCRIPTION
**Pull request by tanim0la**
- implemented `createStrategy` function that stores all the strategies for each curator.
- wrote a test to create and store this strategy:
   ```
  Supply half your cbBTC on Moonwell
  Borrow USDC on Moonwell
  Supply cbBTC + USDC on Aerodrome
   ```
- updated the `assetIn` field in `IStrategy` to an array of addresses to accommodate strategies (like the one above) that require supplying multiple assets.
- added `strategyId` field to the `Strategy` struct in `IStrategy`.

### CODE REVIEW

File: src/curators/strategy.sol [hunk 1]
```diff
@@ -1 +1,64 @@
...
+        uint256 _minDeposit,
+        uint256 _maxTVL,
+        uint256 _performanceFee
+    ) public {
+        bytes32 _strategyId = keccak256(abi.encodePacked(msg.sender, _name));
+        ILiquidStrategy.Strategy memory _strategy = ILiquidStrategy.Strategy({
+            strategyId: _strategyId,
+            curator: msg.sender,
+            name: _name,
+            steps: _steps,
```
Comment (njokuScript) [inline src/curators/strategy.sol:50]: we need to validate the steps before adding it to a strategy. So I'd do something like this:


````
function validateSteps(Step[] memory steps) public view returns (bool) {
        for (uint i = 0; i < steps.length; i++) {
            // Validate protocol address
            require(steps[i].protocol != address(0), "Invalid protocol address");
 
            // Validate assets
            require(steps[i].assetIn != address(0), "Invalid input asset");
            if (steps[i].actionType != ActionType.SUPPLY) {
                require(steps[i].assetOut != address(0), "Invalid output asset");
            }

            // Validate amount ratio
            require(steps[i].amountRatio > 0 && steps[i].amountRatio <= 10000, "Invalid amount ratio"); // Max 100%

            // Validate protocol supports action
            IProtocolIntegration protocol = IProtocolIntegration(steps[i].protocol);
            require(isValidActionForProtocol(
                protocol.protocolType(),
                steps[i].actionType
            ), "Invalid action for protocol type");

            // If not first step, validate asset chain
            if (i > 0) {
                require(
                    steps[i].assetIn == steps[i-1].assetOut,
                    "Asset chain mismatch"
                );
            }
        }
        return true;
    }

    /**
     * @notice Validates if an action is valid for a protocol type
     * @param protocolType Type of protocol
     * @param actionType Type of action
     */
    function isValidActionForProtocol(
        ProtocolType protocolType,
        ActionType actionType
    ) internal pure returns (bool) {
        if (protocolType == ProtocolType.LENDING) {
            return actionType == ActionType.SUPPLY ||
                   actionType == ActionType.WITHDRAW ||
                   actionType == ActionType.BORROW ||
                   actionType == ActionType.REPAY;
        }
        else if (protocolType == ProtocolType.DEX) {
            return actionType == ActionType.SWAP;
        }
        else if (protocolType == ProtocolType.YIELD) {
            return actionType == ActionType.STAKE ||
                   actionType == ActionType.UNSTAKE ||
                   actionType == ActionType.CLAIM;
        }
        return false;
    }

```

I mean, this is not neat, but should help you get the point.

**Review by njokuScript — CHANGES_REQUESTED** [review]
Nice work! Left a couple of comments! @tanim0la

File: src/curators/interface/IStrategy.sol [hunk 2]
```diff
@@ -1,20 +1,22 @@
 pragma solidity ^0.8.20;
 
-import "../../interface/IProtocol.sol";
+import "../../interface/IConnector.sol";
 
-interface ILiquidStrategy is IProtocolIntegration {
+interface ILiquidStrategy is IConnectorIntegration {
```
Comment (njokuScript) [inline src/curators/interface/IStrategy.sol:5]:
```suggestion
interface ILiquidStrategy is IConnector {
```

File: src/curators/strategy.sol [hunk 3]
```diff
@@ -1 +1,147 @@
...
+            }
+            if (steps[i].actionType != IConnectorIntegration.ActionType.SUPPLY) {
+                require(steps[i].assetOut != address(0), "Invalid output asset");
+            }
+
+            // Validate amount ratio
+            require(steps[i].amountRatio > 0 && steps[i].amountRatio <= 10_000, "Invalid amount ratio"); // Max 100%
+
+            // Validate connector supports action
+            IConnectorIntegration connector = IConnectorIntegration(steps[i].connector);
```
Comment (njokuScript) [inline src/curators/strategy.sol:113]:
```suggestion
            IConnector connector = IConnector(steps[i].connector);
```

File: src/curators/strategy.sol [hunk 4]
```diff
@@ -1 +1,147 @@
...
+    }
+
+    /**
+     * @dev Validates if an action is valid for a connector type
+     * @param connectorType Type of connector
+     * @param actionType Type of action
+     */
+    function _isValidActionForConnector(
+        IConnectorIntegration.ConnectorType connectorType,
+        IConnectorIntegration.ActionType actionType
```
Comment (njokuScript) [inline src/curators/strategy.sol:130]:
```suggestion
        IConnector.ConnectorType connectorType,
        IConnector.ActionType actionType
```

File: src/curators/strategy.sol [hunk 5]
```diff
@@ -1 +1,147 @@
...
+    ) internal pure returns (bool) {
+        if (connectorType == IConnectorIntegration.ConnectorType.LENDING) {
+            return actionType == IConnectorIntegration.ActionType.SUPPLY
+                || actionType == IConnectorIntegration.ActionType.WITHDRAW
+                || actionType == IConnectorIntegration.ActionType.BORROW
+                || actionType == IConnectorIntegration.ActionType.REPAY;
+        } else if (connectorType == IConnectorIntegration.ConnectorType.DEX) {
+            return actionType == IConnectorIntegration.ActionType.SWAP;
+        } else if (connectorType == IConnectorIntegration.ConnectorType.YIELD) {
+            return actionType == IConnectorIntegration.ActionType.STAKE
```
Comment (njokuScript) [inline src/curators/strategy.sol:140]: change every reference of IConnectorIntegration to IConnector. Much cleaner

File: src/interface/IConnector.sol [hunk 6]
```diff
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.12;
 
-interface IProtocolIntegration {
+interface IConnectorIntegration {
```
Comment (njokuScript) [inline src/interface/IConnector.sol:4]:
```suggestion
interface IConnector {
```

**Review by njokuScript — CHANGES_REQUESTED** [review]

File: src/curators/strategy.sol [hunk 7]
```diff
@@ -1 +1,145 @@
...
+            // Validate assets
+            for (uint256 j; j < steps[i].assetsIn.length; j++) {
+                require(steps[i].assetsIn[j] != address(0), "Invalid input asset");
+            }
+            if (steps[i].actionType != IConnector.ActionType.SUPPLY) {
+                require(steps[i].assetOut != address(0), "Invalid output asset");
+            }
+
+            // Validate amount ratio
+            require(steps[i].amountRatio > 0 && steps[i].amountRatio <= 10_000, "Invalid amount ratio"); // Max 100%
```
Comment (njokuScript) [inline src/curators/strategy.sol:110]: For this amount ratio, let's think through what happens if it's a one step strategy. Just leaving it as a comment, we can address in the future

**Review by njokuScript — APPROVED** [review]

File: src/curators/strategy.sol (hunk 7 above)
Comment (tanim0la) [reply src/curators/strategy.sol:110]: we can put the `require` statement in an if block that checks if `i` (in this context) is greater than 0 or what do you think?

Comment (njokuScript) [reply src/curators/strategy.sol:110]: > we can put the `require` statement in an if block that checks if `i` (in this context) is greater than 0 or what do you think?

Absolutely! makes sense.

**Pull request merged**